File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
class UnionFind {
    constructor(n){
        this.parent = Array.from({length: n}, (_, i)=>i);
        this.rank = new Array(n).fill(1);
    }
    
    find(x){
        if (this.parent[x]!==x){
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }
    
    union(x, y){
        let rootX = this.find(x);
        let rootY = this.find(y);
        
        if (rootX!==rootY){
            if (this.rank[rootX]>this.rank[rootY]){
                this.parent[rootY] = rootX;
            } else if (this.rank[rootX]<this.rank[rootY]){
                this.parent[rootX] = rootY;
            } else {
                this.parent[rootY] = rootX;
                this.rank[rootX]+=1;
            }
        }
    }
}

function solution(n, wires) {
    let minDiff = Number.MAX_VALUE;
    for (let i=0; i<wires.length; i++){
        const uf = new UnionFind(n);
        for (let j=0; j<wires.length; j++){
            if (j!==i){
                uf.union(wires[j][0], wires[j][1]);
            }
        }
        let team1 = 0; let team2 = 0;
        for (let j=1; j<=n; j++){
            if (uf.find(wires[i][0])===uf.find(j)) team1++;
            if (uf.find(wires[i][1])===uf.find(j)) team2++;
        }
        minDiff = Math.min(minDiff, Math.abs(team1-team2));
    }
    return minDiff;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    static class UnionFind {
        public int[] parent;
        public int[] rank;
        
        UnionFind(int n){
            parent = new int[n];
            rank = new int[n];
            
            for (int i=0; i<n; i++){
                parent[i] = i;
                rank[i] = 1;
            }
        }
        
        public int find(int x){
            if (parent[x]!=x){
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }
        
        public void union(int x, int y){
            int rootX = find(x);
            int rootY = find(y);
            
            if (rootX!=rootY){
                if (rank[rootX]>rank[rootY]) parent[rootY] = rootX;
                else if (rank[rootY]>rank[rootX]) parent[rootX] = rootY;
                else {
                    rank[rootX]++;
                    parent[rootY] = rootX;
                }
            }
        }
    }
    public int solution(int n, int[][] wires) {
        int answer = -1;
        
        int minDiff = Integer.MAX_VALUE;
        for (int i=0; i<wires.length; i++){
            UnionFind uf = new UnionFind(n);
            for (int j=0; j<wires.length; j++){
                if (j!=i){
                    uf.union(wires[j][0]-1, wires[j][1]-1);
                }
            }
            int cnt1 = 0; int cnt2 = 0;
            for (int j=0; j<n; j++){
                if (uf.find(j)==uf.find(wires[i][0]-1)) cnt1++;
                else cnt2++;
            }
            minDiff = Math.min(minDiff, Math.abs(cnt1-cnt2));
        }
        
        return minDiff;
    }
}